vim-patch:8.1.{1418,1425,1832,2124},8.2.{0137, 2340}

ivechan opened the pull request:
To port Vim's win_execute() function.  https://github.com/neovim/neovim/issues/10822
extra tests should be added into tests/functinal/ui/float_spec.lua

bfredl commented:
nice! We should also add `nvim_win_call` for lua in analogy with `nvim_buf_call`, but I will do that as a follow-up PR.

bfredl commented on src/nvim/window.c:
```diff
@@ -6528,6 +6533,16 @@ static win_T *get_snapshot_focus(int idx)
+}
+
+// As switch_win() but without blocking autocommands.
+int switch_win_noblock(win_T **save_curwin,
```
we should also use the "noblock" variants in `nvim_win_set_buf`

ivechan replied:
done! and add a testcase in test/functional/autocmd/window_spec.lua .

bfredl commented:
I think this should be in 0.5 as it will help fix some bugs :]

ivechan commented:
@bfredl I think current legacy tests in this PR is enough.
But I can't figure out why CI / windows (MSVC_32) failed because the error on MSVC_32 seems to be not related to this PR.

ivechan commented:
This PR is ready to be merged if you agreed.

ivechan committed:
api/window: use the "noblock" variants in nvim_win_set_buf

after commit 92c6383cdca977("vim-patch:8.1.1425: win_execute() does not set window pointers properly"),
nvim_win_set_buf can use switch_win_noblock and restore_win_noblock.
It makes nvim_win_set_buf don't block autocmds so that it will be more
"without side-effects" as said in help text.

Signed-off-by: jing <lhchenjw@gmail.com>

ivechan commented:
now autocmd can work as expected when calling nvim_win_set_buf.
I added a testcase to make sure it work fine.

janlazo commented:
Commit format on some commits and the PR title are wrong. It's causing the PR to not be auto-labeled with the `vim-patch` label.

ivechan commented:
> Commit format on some commits and the PR title are wrong. It's causing the PR to not be auto-labeled with the `vim-patch` label.

I apologize it.
I have corrected the format and resubmitted, please review it agian :)

janlazo commented on runtime/doc/eval.txt:
```diff
@@ -9262,6 +9262,12 @@ wildmenumode()					*wildmenumode()*
 <
 		(Note, this needs the 'wildcharm' option set appropriately).
 
+win_execute({id}, {command} [, {silent}])		*win_execute()*
```
This is not the only change in this file.

janlazo commented on src/nvim/eval/funcs.c:
```diff
@@ -1982,6 +1986,37 @@ static void f_execute(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+/// "execute()" function
+static void f_execute(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+{
+    execute_common(argvars, rettv, fptr, 0);
```
2-space indent.

janlazo commented on src/nvim/testdir/test_execute_func.vim:
```diff
@@ -82,3 +84,43 @@ func Test_execute_not_silent()
+  call win_gotoid(thiswin)
+  let line = win_execute(otherwin, 'echo getline(1)')
+  call assert_match('the new window', line)
+
```
https://github.com/vim/vim/commit/868b7b6712ea4f2232eeeae18c5cbbbddf2ee84d#diff-2fa6230b9e8601cd9e1fae2257f1c95f95e307ede63335c924646f90582eee34R91-R100 is missing.

Do not remove Vimscript code in new-style Vim tests that should be skipped in Neovim.

ivechan commented:
WIP, many issue need to be address.

ivechan committed:
vim-patch:8.1.1418: win_execute() is not implemented yet

Problem:    Win_execute() is not implemented yet.
Solution:   Implement it.
https://github.com/vim/vim/commit/868b7b6712ea4f2232eeeae18c5cbbbddf2ee84d

ivechan committed:
vim-patch:8.1.1425: win_execute() does not set window pointers properly

Problem:    Win_execute() does not set window pointers properly.
Solution:   Use switch_win_noblock().  Also execute autocommands in a popup
            window.
https://github.com/vim/vim/commit/89adc3a1371d211f7766f3dbc0975ecb2f862327

ivechan committed:
vim-patch:8.1.1440: win_execute() test fails

Problem:    Win_execute() test fails.
Solution:   Adjust the expected error number.  Move to popup test.
https://github.com/vim/vim/commit/2d247849ce612050ba1085df806746b23be1f0a3

ivechan committed:
vim-patch:8.1.1832: win_execute() does not work in other tab

Problem:    Win_execute() does not work in other tab. (Rick Howe)
Solution:   Take care of the tab. (closes vim/vim#4792)
https://github.com/vim/vim/commit/820680b9ff1de8699156c7b060f97e5c0b87ad15

ivechan committed:
vim-patch:8.1.2124: ruler is not updated if win_execute() moves cursor

Problem:    Ruler is not updated if win_execute() moves cursor.
Solution:   Update the status line. (closes vim/vim#5022)
https://github.com/vim/vim/commit/345f28df5482cd35f5fa74b06443376379f113b0

ivechan committed:
vim-patch:8.2.0137: crash when using win_execute() from a new tab

Problem:    Crash when using win_execute() from a new tab.
Solution:   Set the tp_*win pointers. (Ozaki Kiichi, closes vim/vim#5512)
https://github.com/vim/vim/commit/a44b3eeafa57d4904a3de86b132008b93404f0fd

ivechan committed:
vim-patch:8.2.2340: win_execute() unexpectedly returns number zero when failing

Problem:    win_execute() unexpectedly returns number zero when failing.
Solution:   Return an empty string. (closes vim/vim#7665)
https://github.com/vim/vim/commit/37487e16da7877129edee8d11b9b7f5c8df312c6

ivechan commented:
I re-finished this PR
And this time I was  careful to avoid unexpected inconsistencies with the original vim patches.

text_prop, popup and vim9 related changes are not introduced to src/

janlazo commented:
Adding back to 0.5.0 if it's ready because it's mentioned in  https://github.com/neovim/neovim/pull/14424. I shouldn't have done that but I don't want to amend any commits there.

janlazo commented on src/nvim/window.c:
```diff
@@ -6359,6 +6370,13 @@ int switch_win(win_T **save_curwin, tabpage_T **save_curtab, win_T *win, tabpage
+  unblock_autocmds();
+}
+
+void restore_win_noblock(win_T *save_curwin, tabpage_T *save_curtab,
```
Function description missing.

janlazo commented on src/nvim/window.c:
```diff
@@ -6335,6 +6340,12 @@ static win_T *get_snapshot_focus(int idx)
+  return switch_win_noblock(save_curwin, save_curtab, win, tp, no_display);
+}
+
+int switch_win_noblock(win_T **save_curwin, tabpage_T **save_curtab,
```
Function description missing.

janlazo commented:
No lua tests? Vim's popup code is not ported so some test cases here don't run in the CI.

janlazo reviewed (changes requested):
LGTM aside from my comments.

ivechan commented:
> No lua tests? Vim's popup code is not ported so some test cases here don't run in the CI.
What test cases do you think need to be added?
I plan to port the testcases of .vim to .lua and replace popup related testcases with float window testcases。

janlazo commented:
> What test cases do you think need to be added?

Any tests you already have for `tests/functinal/ui/float_spec.lua`.
Optionally, any Vim screendump tests that you ported to a Neovim screen test.

ivechan commented:
The function descriptions were added.
And I just added two basic testcases in `test/functional/ui/float_spec.lua`

janlazo reviewed (approved):
LGTM

clason commented:
Just to note this here: this PR broke telescope; trying to open a finder throws
```
E5108: Error executing lua .../packer/start/plenary.nvim/lua/plenary/window/border.lua:138: Vim(autocmd):E680: <buffer=6>: invalid buffer number
```
If anybody here has a quick idea what needs to be changed in telescope to make it work with the changed code, that would be helpful.

kevinhwang91 commented:
After the patch, `nvim_open_win` call `nvim_win_set_buf` and throw error. Please give a correct way for using `nvim_open_win`.

ray-x commented:
I am not 100% sure what this means to autocommand
The fix for telescope is removing "BufLeave" from autocommand.  
But what happens if the plugin relay on "BufLeave"? Is there a new API to set autocommand?

tomtomjhj commented:
I don't quite understand https://github.com/neovim/neovim/pull/13664/commits/1def3d1542d6a65f057e743faea39a760b50db87. This commit makes `nvim_win_set_buf` not block autocmds so that it's more "without side-effects", but I think allowing autocmds results in more side-effects. Has anyone tried reverting this commit? My setup is not affected by this PR, so I can't test myself.

ddickstein commented:
@tomtomjhj I may be missing some context here but I think that when a user expects to run certain logic every time Foo happens (expressed by registering an autocmd) and then you trigger Foo but prevent that logic from running (noautocmd), you are circumventing the user's intention / doing something surprising. Maybe that's what the author means by "without side-effects" -- without special behavior. But in the literal sense it's definitely more "effectful" to run logic on events than to not run logic on events.

janlazo commented:
Created https://github.com/neovim/neovim/pull/14531 to revert the breaking changes if there's no fix within this month.

bfredl commented:
I would rather add a `noautocmd=true` flag to `nvim_open_win()`. Running Enter autocmds for the specific buffer-window pair is a __main__ effect of showing that buffer in the window, and needed to support filetypes, cmd buffers and a lot of other stuff (the side effects we avoid is unnecessary winenter/leave autocmds of _other_ windows which a naive user-space re-implementation of `nvim_win_set_buf` would could cause)

ivechan commented:
What @ddickstein said explains what I meant by  "without side-effects".


> Is there a missing +popupwin patch that should be ported to Neovim's floating window?

I will check it.

> I would rather add a noautocmd=true flag to nvim_open_win().

Are there any scenarios where we would expect nvim_win_set_buf to trigger autocmd? 
If not, I think it might be better to just revert this commit

bfredl commented:
> Are there any scenarios where we would expect nvim_win_set_buf to trigger autocmd?
If not, I think it might be better to just revert this commit

The same scenarios where `:edit` and `:buffer` is expected to trigger autocmds, of which I listed some above. Most importantly filetype (and bufreadcmd), which is now only supported in `nvim_open_win` through hacks.  Editing a buffer in a floating window should generally be expected to behave the same as using another window, and it is much simpler and reliable to disable autocmds when we don't want them that faking them where they got missing.
